### DISCUSSION TITLE
Bizarre chocolatey Doxygen issue after Windows Server 2022 20220619.1 image update

### DISCUSSION THREAD

**Discussion by oblivioncth** [community/community#26278] [Actions]
<p>I don’t really expect help with this given how nebulous the problem is, but I figured I’d put it out there just in case since this will likely take a problematically long time to figure out.</p>
<p>For one of my projects I use actions on GitHub hosted runners to handle CI/CD. This project makes use of Doxygen, which does not come by default with the Windows Server 2022 image.</p>
<p>So, with that in mind, here is a simplified version of the job that performs the main build for this project</p>
<pre><code class="lang-auto">jobs:
  build-qx:
    name: Build Qx - Windows (Debug/Release)
    strategy:
      matrix:
        qt_linkage: [shared, static]
    runs-on: windows-latest
    env:
      vs_dir: C:/Program Files/Microsoft Visual Studio/2022/Enterprise
      cmake_gen: Ninja Multi-Config
    outputs:
      qt_shared_artifact_name: ${{ steps.get_artifact_name.outputs.qt_shared_artifact_name }}
      qt_static_artifact_name: ${{ steps.get_artifact_name.outputs.qt_static_artifact_name }}
    steps:
    (...)
    - name: Cache Qt Build
      id: cache-qt
      uses: actions/cache@v3
      with:
        path: ${{ env.qt_install_dir }}
        key: ${{ env.qt_install_name }}
    (...)
    - name: Install Doxygen &amp; Graphviz
      run: |
        choco install doxygen.install --version 1.9.4
        choco install graphviz
    - name: Checkout Qx
      uses: actions/checkout@v3
      with:
        path: ${{ env.qx_src_suffix }}
        fetch-depth: 0 # Required for verbose versioning to work correctly
    - name: Build/Install Qx
      working-directory: ${{ env.qx_src_dir }}
      shell: cmd
      run: |
        echo "Setup C++ Build Environment..."
        CALL "${{ env.vs_dir }}\Common7\Tools\VsDevCmd.bat" -arch=amd64
        echo "Configure CMake using Qt wrapper..."
        CALL "${{ env.qt_cmake }}" -G "${{ env.cmake_gen }}" -S "${{ env.qx_src_dir}}" -B "${{ env.qx_build_dir }}" -D QX_DOCS_TARGET=ON
        echo "Changing to build directory..."
        cd "%qx_build_dir%"
        echo "Building Qx debug..."
        cmake --build . --target all --config Debug
        echo "Building Qx release..."
        cmake --build . --target all --config Release
        echo "Building Qx docs..."
        cmake --build . --target qx_docs --config Release
        echo "Installing Qx Debug"
        cmake --build . --target install --config Debug
        echo "Installing Qx Release/Docs..."
        cmake --build . --target install --config Release
        echo "Packaging Qx..."
        cpack -C "Debug;Release"
        echo "Build complete."
    (...)
</code></pre>
<p>This <em>was</em> working fine. What started happening is that, at first intermittently, documentation would fail to build on the Windows Server 2022 hosts. Checking the logs, when CMake ran I see that instead of the usual:<br>
<code>Found Doxygen: C:/Program Files/doxygen/bin/doxygen.exe (found version "1.9.4...</code></p>
<p>I see:<br>
<code>Found Doxygen: C:/ProgramData/chocolatey/bin/doxygen.exe (found version "1.9.1...</code></p>
<p>and then the build later fails because of Doxygen features I use that were added after 1.9.1.</p>
<p>I first assumed that something must have happened with the chocolatey package, but it hasn’t been touched and the logs for that Doxygen installation step are nearly identical (other than run-to-run variance) between the runs that see the right version vs. the wrong version. It shows the setup executable for 1.9.4 being ran in both cases. So all-in-all I’m not sure what’s up with the path change.</p>
<p>Needless to say I was thoroughly confused, but decided to focus on the fact that at first this problem started happening sporadically, yet as of now happens every time. <strong>I even have a single <a href="https://github.com/oblivioncth/Qx/actions/runs/2545870245" rel="noopener nofollow ugc">run</a> where the first path of the Windows build matrix found the correct Doxygen version and the second path found the wrong Doxygen version.</strong></p>
<p>After some cross-referencing, I was able to figure out a consistent demarcation point:</p>
<p>It seems that these mixed success runs happened to occur during the rollout of the <code>Windows Server 2022 20220619.1</code> image, as some of the runs were executed on a host with <code>20220614.0</code>, while others were on a host with <code>20220619.1</code> (and very recently a few with <code>20220626.1</code>).</p>
<p><strong>100% of the runs where CMake found the correct version of Doxygen (1.9.4) occurred on a runner with <code>20220614.0</code>, while 100% of the runs where the wrong version of Doxygen (1.9.1) that seemingly pops into existence out of nowhere is found occurred on a runner with <code>20220619.1</code> or <code>20220626.1</code>.</strong></p>
<p>My workflow has not been altered at all during any of this, especially evident by the fact this correlation of Windows Server 2022 image version was observed during a single matrix run as noted earlier.</p>
<p>Figuring that some related tool must have been changed with the <code>20220619.1</code> update, I went over to compare the two ReadMe files and found… nothing.</p>
<aside class="onebox githubcommit" data-onebox-src="https://github.com/actions/virtual-environments/commit/96c2ea89565b32d23595ce34732dc7539d67166e">
  <header class="source">

      <a href="https://github.com/actions/virtual-environments/commit/96c2ea89565b32d23595ce34732dc7539d67166e" target="_blank" rel="noopener nofollow ugc">github.com/actions/virtual-environments</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/virtual-environments/commit/96c2ea89565b32d23595ce34732dc7539d67166e" target="_blank" rel="noopener nofollow ugc">Updating readme file for win22 version 20220619.1 (#5783)</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2022-06-23" data-time="14:37:16" data-timezone="UTC">02:37PM - 23 Jun 22 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/apps/github-actions" target="_blank" rel="noopener nofollow ugc">
          <img alt="github-actions[bot]" src="https://user-images.githubusercontent.com/24661585/181096619-22ed8fc9-0783-4235-8744-449a15826ee0.png" class="onebox-avatar-inline" width="20" height="20">
          github-actions[bot]
        </a>
      </div>

      <div class="lines" title="changed 1 files with 18 additions and 16 deletions">
        <a href="https://github.com/actions/virtual-environments/commit/96c2ea89565b32d23595ce34732dc7539d67166e" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+18</span>
          <span class="removed">-16</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">Co-authored-by: Image generation service account &lt;no-reply@microsoft.com&gt;
Co-aut<span class="show-more-container"><a href="https://github.com/actions/virtual-environments/commit/96c2ea89565b32d23595ce34732dc7539d67166e" target="_blank" rel="noopener nofollow ugc" class="show-more">…</a></span><span class="excerpt hidden">hored-by: Actions service account &lt;no-reply@github.com&gt;</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Chocolatey wasn’t updated, CMake wasn’t updated, Powershell wasn’t updated, even the build of Windows Server 2022 wasn’t updated.</p>
<p>The only things that were updated that are even within the same universe of relevance were:</p>
<ul>
<li>vcpkg</li>
<li>Microsoft.VisualStudio.Component.Windows11SDK.22621</li>
</ul>
<p>But vcpkg should have no bearing on any of this unless its somehow used indirectly by something else, and I really doubt that a new Windows SDK version would cause such an issue (I’d hope), and I don’t even think it’s used by default.</p>
<p>I changed my <code>find_package(Doxygen...)</code> call so that it specifically requires 1.9.4 in hopes that for whatever bizarre reason two version of Doxygen were being installed and that I could at least just ignore the wrong one, but unfortunately that didn’t work. It seems that after the image update, only 1.9.1 gets installed as I now get:<br>
<code>Could NOT find Doxygen: Found unsuitable version "1.9.1", but required is at least "1.9.4" (found C:/ProgramData/chocolatey/bin/doxygen.exe...</code></p>
<p>I’m fairly lost here. This basically just doesn’t make sense.</p>
<p>I don’t supposed anyone has a clue as to what’s going on.</p>

**Comment by oblivioncth** [comment]
<p>Ok, I still need to dig much deeper, but I’m onto something. Mingw was updated as part of that image update, quite a number of versions I might add, and it seems to be installed via chocolatey actually.</p>
<p>By default, even before Doxygen is installed via my script, there exists a choco shim at <code>C:\ProgramData\chocolatey\bin\doxygen.exe</code> that points to <code>C:\ProgramData\chocolatey\lib\mingw\tools\install\mingw64\bin\doxygen</code>. Given that its the shim thats being found, it seems that the version of mingw included in the image (11.2.0) contains Doxygen 1.9.1</p>
<p>What’s still strange is why the install of the Doxygen package still can’t be picked up by CMake. Even if theres some issue where the custom install is unable to overwrite the existing shim, it shouldn’t matter as before this image version FindDoxygen had no issue locating it at <code>C:\Program Files\doxygen</code>, where it was, and still is installed.</p>
<p><strong>EDIT:</strong><br>
I can confirm that Mingw 8.1.0, which was on the old Windows Server 2022 image, did not include a build of doxygen, while Mingw 11.2.0, which is on the newer image, does.</p>
<p>Hopefully I can figure out why it breaks detection of the separate Doxygen package.</p>

**Comment by oblivioncth** [comment, marked as answer]
<p>Alright I cracked it.</p>
<p>There are several things going on here, but ultimately it is mostly a CMake issue.</p>
<p>When calling <code>find_package(Doxygen...</code> in CMake, the build-in FindDoxygen.cmake module is called, which in turn attempts to locate Doxygen using the following:</p>
<pre><code class="lang-auto">find_program(
        DOXYGEN_EXECUTABLE
        NAMES doxygen
        PATHS
            "[HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\doxygen_is1;Inno Setup: App Path]/bin"
            /Applications/Doxygen.app/Contents/Resources
            /Applications/Doxygen.app/Contents/MacOS
            /Applications/Utilities/Doxygen.app/Contents/Resources
            /Applications/Utilities/Doxygen.app/Contents/MacOS
        DOC "Doxygen documentation generation tool (http://www.doxygen.org)"
)
</code></pre>
<p>The following is an extremely simplified overview of the search process for <code>find_program</code>, with only the aspects that come in to play:</p>
<ol>
<li>(Since 3.12) If the call to <code>find_program</code> is part of a <code>find_package(&lt;PackageName&gt;)</code> call, search all prefixes in <code>PackageName_ROOT</code> and any <code>bin</code> subdirectories within those</li>
<li>Search the prefixes (and <code>bin</code> folders within them) of <code>CMAKE_PREFIX_PATH</code>
</li>
<li>Search prefixes of <code>CMAKE_PROGRAM_PATH</code> and <code>CMAKE_APPBUNDLE_PATH</code></li>
<li>Search standard system environment variables</li>
<li>Search the paths provided by the PATHS option</li>
</ol>
<p>Unlike <code>find_package()</code>, <code>find_program()</code> cannot take into account version requirements, so even though the <code>find_package()</code> call that wraps it does, the function will simply report the location of the first exe match it finds every time (in fact it only checks once and caches the result) and <code>find_package()</code> will keep seeing that version.</p>
<p>Basically, CMake does not handle having multiple versions of Doxygen installed gracefully, even when a version requirement should theoretically disambiguate them.</p>
<p>Well, due to the way the above call is structured, the standalone version of Doxygen on chocolatey is found on the system via its uninstall entry in the registry, as specified under the PATHS option. The shim of the version included with mingw on chocolatey is located via the system PATH environment variable that chocolatey adds by default: <code>C:\ProgramData\chocolatey\bin</code>.</p>
<p>Unfortunately, as can been see in the above search order, the system PATH variable has a higher precedence than the PATHS option of the FindDoxygen command, so by default, unassisted calls to <code>find_package(Doxygen...</code> will always only find the mingw version in this situation, causing the problem I have.</p>
<p>While in the long run there has to be a better way (which unfortunately will likely require CMake to implement FindDoxygen.cmake in a way that can handle checking multiple installs for a specific version), currently this can be worked around by:</p>
<ol>
<li>Adding the main Doxygen install path to <code>Doxygen_ROOT</code> (though I can’t seem to get this to work)</li>
<li>Adding the main Doxygen install path to <code>CMAKE_PREFIX_PATH</code>
</li>
<li>Adding the main Doxygen install path to <code>CMAKE_PROGRAM_PATH</code>
</li>
<li>Writing your own FindDoxygen.cmake, which likely would be nearly the same as the official one, but call <code>find_program</code> with different arguments to change the search order such that the standalone Doxygen install is found first</li>
</ol>
<p>Personally I find it annoying that the chocolatey distro of mingw comes with a build of Doxygen since a newer one is readily available within chocolatey already, but ultimately that isn’t the root issue.</p>
<p>In the end this wasn’t an Actions problem at all, but I have to imagine at least one other person was affected by this change.</p>

**Reply by sbzrd** [reply]
Thanks for sharing - was struggling with this issue myself all morning after updating pipeline vm image to windows-2022. 

Upgrading mingw installation via choco to 12.2.0 also upgrades the mingw shipped doxygen to 1.9.5. However this is a little time consuming (>2mins in my pipeline).

I ended up downgrading vm image to windows-2019 and using choco to just install the doxygen version I need. Since this is faster and a little more "controlled". And since I don't really need anything from windows 2022 image.

Anyways, I probably could have spend some more hours debugging this if I hadn't come across your post :-)

**Reply by oblivioncth** [reply]
Hey, glad it at least helped someone else!

I didn't end up mentioning here two important things.

1) I forget what I was doing wrong the first time but I can confirm that setting Doxygen_ROOT to the path of the Doxygen install you want to prioritize does work, as that's how all of my workflows currently deal with the issue.

2) Sometime after this I actually submitted a PR to CMake with a modified FindDoxygen.cmake that fixes the issue so that the version number requested in 'find_package(Doxygen ...)' is actually honored. It was merged in for CMake 3.25.0 which is currently on RC4, and so should be available real soon.

GitHub seems to keep the version of CMake on their runners up-to-date so assuming they update their images to 3.25.0 once it's out this won't be an issue any longer.
